.pre-commit-config.yaml

```diff
@@ -1,8 +1,92 @@
+    rev: v5.0.0
     hooks:
-      - id: black
-        language_version: python3.12
-
-
+      - id: check-added-large-files
+        args: [— maxkb=1000]
+        name: Check for added large files
+      - id: check-ast
+        name: Check Python files for parse errors
+      - id: check-case-conflict
+        name: Check files for case conflicts on case-insensitive filesystems
+      - id: check-docstring-first
+        name: Check that no code is placed prior to docstrings
+      - id: check-executables-have-shebangs
+        name: Check that executables have shebangs
+      - id: check-illegal-windows-names
+        name: Check for illegal Windows filenames
+      - id: check-json
+        name: Check valid JSON syntax
+      - id: check-merge-conflict
+        name: Check for files that contain merge conflict strings
+      - id: check-shebang-scripts-are-executable
+        name: Check that shebang scripts are executable
+      - id: check-toml
+        name: Check valid TOML syntax
+      - id: check-xml
+        name: Check valid XML syntax
+      - id: check-yaml
+        name: Check valid YAML syntax
+      - id: end-of-file-fixer
+        name: Check for newline at end of file
+      - id: debug-statements
+        name: Check for debugger imports
+      - id: no-commit-to-branch
+        name: Check to not commit to main
+      - id: trailing-whitespace
+        name: Check for trailing whitespaces
+  - repo: https://github.com/PyCQA/bandit
+    rev: 1.8.0
+    hooks:
+      - id: bandit
+        name: Run bandit (security linter for python)
+        args: ["-c", "pyproject.toml", "-r", "-ll", "meshpy/"]
+        additional_dependencies: ["bandit[toml]"]
+  - repo: https://github.com/pre-commit/mirrors-clang-format
+    rev: v19.1.5
+    hooks:
+      - id: clang-format
+        name: Run clang-format (C/C++ formatter)
+  - repo: https://github.com/PyCQA/docformatter
+    rev: "eb1df347edd128b30cd3368dddc3aa65edcfac38" # change back to version once bug is fixed in latest version to be compatible with pre-commit
+    hooks:
+      - id: docformatter
+        name: Run docformatter (formatter for docstrings)
+        args: ["--in-place", "--config=./pyproject.toml", "-r", "meshpy/"]
+  - repo: https://github.com/econchick/interrogate
+    rev: 1.7.0
+    hooks:
+      - id: interrogate
+        name: Run interrogate (linter for docstrings)
+        args: [--fail-under=95, --ignore-init-module, --style=google, -vv, meshpy/]
+  - repo: https://github.com/pre-commit/mirrors-mypy
+    rev: v1.13.0
+    hooks:
+      - id: mypy
+        name: Run mypy (static type checker for python)
+        args: ["--install-types", "--non-interactive", "--ignore-missing-imports", "--follow-imports=silent"]
+  - repo: https://github.com/asmeurer/removestar
+    rev: "1.5.2"
+    hooks:
+      - id: removestar
+        name: Run removestar (remove wildcard imports)
+  - repo: https://github.com/astral-sh/ruff-pre-commit
```

Comment:
Just so I understand, this is compatible with the `black` code style?

Reply:
Yes - ruff more or less applies the black formatting style, per their docs:

> The initial goal of the Ruff formatter is not to innovate on code style, but rather, to innovate on performance, and provide a unified toolchain across Ruff's linter, formatter, and any and all future tools.
> 
> As such, the formatter is designed as a drop-in replacement for [Black](https://github.com/psf/black), but with an excessive focus on performance and direct integration with Ruff. Given Black's popularity within the Python ecosystem, targeting Black compatibility ensures that formatter adoption is minimally disruptive for the vast majority of projects.
> 
> Specifically, the formatter is intended to emit near-identical output when run over existing Black-formatted code. When run over extensive Black-formatted projects like Django and Zulip, > 99.9% of lines are formatted identically. (See: [Black compatibility](https://docs.astral.sh/ruff/formatter/#black-compatibility).)

Given this focus on Black compatibility, the formatter thus adheres to [Black's (stable) code style](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html), which aims for "consistency, generality, readability and reducing git diffs"